### PR TITLE
Fix #5497: Implement Ads auth refresh flow

### LINKED ISSUE #5497: [iOS] New Ads Auth Refresh Flow
## Description

See: https://github.com/brave/internal/issues/881 

@kylehickinson

### PR DESCRIPTION
**Pull request by kylehickinson**
## Summary of Changes

This pull request fixes #5497 

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:
- See internal issue

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/DeviceCheck/DeviceCheck.swift [hunk 1]
```diff
@@ -344,58 +310,64 @@ private extension DeviceCheckClient {
       case .setAttestation(let nonce, _):
         let nonce = nonce.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? nonce
         return URL(string: "/v1/devicecheck/attestations/\(nonce)", relativeTo: client.baseURL)
+      case .solveCaptcha(let paymentId, let captchaId, _):
+        return URL(string: "/v3/captcha/solution/\(paymentId)/\(captchaId)", relativeTo: client.baseURL)
```
Comment (iccub) [inline Client/DeviceCheck/DeviceCheck.swift:316]: I wonder if we need to add any safeguards or sanitization here like above with `nonce` percent encoding?

File: Client/DeviceCheck/DeviceCheck.swift [hunk 2]
```diff
@@ -420,6 +392,10 @@ private extension DeviceCheckClient {
     case .setAttestation(_, let parameters):
       request.httpBody = try JSONEncoder().encode(parameters)
       request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+      
+    case .solveCaptcha(_, _, let nonce):
+      request.httpBody = try JSONSerialization.data(withJSONObject: ["solution": nonce])
```
Comment (iccub) [inline Client/DeviceCheck/DeviceCheck.swift:399]: this nonce does not need percent encoding? i guess it's fine in http

File: Client/Extensions/Rewards/BraveLedgerExtensions.swift [hunk 3]
```diff
@@ -144,82 +144,84 @@ extension BraveLedger {
...
-    let group = DispatchGroup()
     if !DeviceCheckClient.isDeviceEnrolled() {
-      group.enter()
-      setupDeviceCheckEnrollment(deviceCheck) {
-        if !DeviceCheckClient.isDeviceEnrolled() {
-          DispatchQueue.main.async {
-            completion(false)
-          }
-          return
+      let didEnrol: Bool = await withCheckedContinuation { c in
```
Comment (iccub) [inline Client/Extensions/Rewards/BraveLedgerExtensions.swift]: typo, `didEnroll`

File: Client/Extensions/Rewards/BraveLedgerExtensions.swift [hunk 4]
```diff
@@ -144,82 +144,84 @@ extension BraveLedger {
...
+          c.resume(returning: false)
+        }
+      }
+    }
+  }
+  
+  func solveAdaptiveCaptcha(paymentId: String, captchaId: String) async throws {
+    let deviceCheck = DeviceCheckClient(environment: BraveLedger.environment)
+    if !DeviceCheckClient.isDeviceEnrolled() {
+      let didEnrol: Bool = await withCheckedContinuation { c in
```
Comment (iccub) [inline Client/Extensions/Rewards/BraveLedgerExtensions.swift]: same typo, unless i do not know English :D

File: Client/Extensions/Rewards/BraveLedgerExtensions.swift (hunk 3 above)
Comment (kylehickinson) [reply Client/Extensions/Rewards/BraveLedgerExtensions.swift]: It is one of those weird spellings in US vs the rest of the world 😆. For instance if I type "Enroll", macOS attempts to autocorrect it to "Enrol"

File: Client/DeviceCheck/DeviceCheck.swift (hunk 2 above)
Comment (kylehickinson) [reply Client/DeviceCheck/DeviceCheck.swift:399]: AFAIK it does not need to be percent encoded as its just a UUID string

File: Client/DeviceCheck/DeviceCheck.swift (hunk 1 above)
Comment (kylehickinson) [reply Client/DeviceCheck/DeviceCheck.swift:316]: I can add some protections for the path

**Review by iccub — APPROVED** [review]
Going to hit approve as my questions are fine to be skipped and easy to resolve if needed, feel free to resolve my comments if you decide to act on them.

Up to you whether we would like a sec review for it or not

File: Client/Extensions/Rewards/BraveLedgerExtensions.swift (hunk 3 above)
Comment (kylehickinson) [reply Client/Extensions/Rewards/BraveLedgerExtensions.swift]: Changed them to `enroll` spelling as thats what we've been using across the rest of the `DeviceCheckClient`

File: Client/DeviceCheck/DeviceCheck.swift (hunk 1 above)
Comment (kylehickinson) [reply Client/DeviceCheck/DeviceCheck.swift:316]: Added similar protections as nonce

**Pull request merged**